File: src/main/java/com/freya02/botcommands/parameters/ComponentParameterResolver.java
package com.freya02.botcommands.parameters;

import net.dv8tion.jda.api.events.interaction.GenericComponentInteractionCreateEvent;
import org.jetbrains.annotations.Nullable;

/**
 * Interface which indicates this class can resolve parameters for buttons commands
 */
public interface ComponentParameterResolver {
	@Nullable
	Object resolve(GenericComponentInteractionCreateEvent event, String arg);
}

File: src/main/kotlin/io/github/freya022/botcommands/api/parameters/ComponentParameterResolver.kt
package io.github.freya022.botcommands.api.parameters

import io.github.freya022.botcommands.internal.components.ComponentDescriptor
import net.dv8tion.jda.api.events.interaction.component.GenericComponentInteractionCreateEvent

/**
 * Interface which indicates this class can resolve parameters for component interactions.
 */
interface ComponentParameterResolver<T, R> where T : ParameterResolver<T, R>,
                                                 T : ComponentParameterResolver<T, R> {
    /**
     * Returns a resolved object from this component interaction
     *
     * @param descriptor The component description of the component being executed
     * @param event      The event of this component interaction
     * @return The resolved option mapping
     */
    fun resolve(descriptor: ComponentDescriptor, event: GenericComponentInteractionCreateEvent, arg: String): R? =
        throw NotImplementedError("${this.javaClass.simpleName} must implement the 'resolve' or 'resolveSuspend' method")

    @JvmSynthetic
    suspend fun resolveSuspend(
        descriptor: ComponentDescriptor,
        event: GenericComponentInteractionCreateEvent,
        arg: String
    ) = resolve(descriptor, event, arg)
}